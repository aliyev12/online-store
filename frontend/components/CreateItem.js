import React, { useState } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
      title
    }
  }
`;

const CreateItem = () => {
  const [title, setTitle] = useState('Item Title');
  const [description, setDescription] = useState('Item Description');
  const [image, setImage] = useState('');
  const [largeImage, setLargeImage] = useState('');
  const [price, setPrice] = useState(9998);

  const uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'greatfits');
    // https://res.cloudinary.com/greatfits/image/upload/v1564874789/sample.jpg
    const res = await fetch(`https://api.cloudinary.com/v1_1/greatfits/image/upload`, {
      method: 'POST',
      body: data
    });
    const file = await res.json();
    setImage(file.secure_url);
    setLargeImage(file.eager[0].secure_url);
  }

  return (
    <Mutation
      mutation={CREATE_ITEM_MUTATION}
      variables={{
        title,
        description,
        image,
        largeImage,
        price
      }}
    >
      {(createItem, { loading, error }) => (
        <Form
          onSubmit={ async e => {
            // Stop the form from submitting
            e.preventDefault();
            // Call the mutation
            const res = await createItem();
            // Change them to the single item page
            Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id }
            });
          }}
          method="post"
        >
        <Error error={error}/>
          <fieldset disabled={loading} aria-busy={loading}>
            <label htmlFor="file">
              Image
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                required
                onChange={uploadFile}
              />
              {image && <img src={image} alt="Upload Preview" width="200"/>}
            </label>
            <label htmlFor="title">
              Title
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Title"
                required
                value={title}
                onChange={e => setTitle(e.target.value)}
              />
            </label>
            <label htmlFor="price">
              Price
              <input
                type="number"
                id="price"
                name="price"
                placeholder="Price"
                required
                value={price}
                onChange={e => setPrice(e.target.value)}
              />
            </label>
            <label htmlFor="title">
              Description
              <textarea
                id="description"
                name="description"
                placeholder="Enter a Description"
                required
                value={description}
                onChange={e => setDescription(e.target.value)}
              />
            </label>
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
    </Mutation>
  );
};

export default CreateItem;
