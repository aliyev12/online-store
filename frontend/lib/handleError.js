export default (error) => {
    console.log(`Error from handleError.js: `, error);
    // throw new Error(`Something went wrong. Check console for more details`);
    return `Error: ${error && error.message ? error.message : 'No error message'}`;
};