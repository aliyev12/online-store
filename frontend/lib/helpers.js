// Check deeply nested props
/* Example:
const user = {
  posts: [
    {title: 'Title', comment: 'Comment'}
  ]
}
idx(['posts', '0', 'title'], user) // returns 'Title'
idx(['posts', '2', 'title'], user) // returns 'null' because there is no index 2 in posts
*/
export const idx = (p, o) => p.reduce((xs, x) => (xs && xs[x]) ? xs[x] : null, o);