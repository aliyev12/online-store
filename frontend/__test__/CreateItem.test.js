import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import Router from 'next/router';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

const testImage = 'https://test.com/test-image.jpg';

// Mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: testImage,
    eager: [{ secure_url: testImage }]
  })
});

describe('<CreateItem/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const imageInput = wrapper.find('input[type="file"]');
    imageInput.simulate('change', { target: { files: ['test-image.jpg'] } });
    await wait();
    wrapper.update();
    const image = wrapper.find('img');
    expect(image.props().src).toEqual(testImage);
    expect(global.fetch).toHaveBeenCalled();
    global.fetch.mockReset();
  });

  it('creates an item when the form is submitted', async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            description: item.description,
            image: '',
            largeImage: '',
            price: item.price
          }
        },
        result: {
          data: {
            createItem: {
              ...fakeItem(),
              __typeName: 'Item'
            }
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    );
    wrapper
      .find('#title')
      .simulate('change', { target: { value: item.title, name: 'title' } });
    wrapper.find('#price').simulate('change', {
      target: { value: item.price, name: 'price', type: 'number' }
    });
    wrapper.find('#description').simulate('change', {
      target: { value: item.description, name: 'description' }
    });
    // Mock the router
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/item',
      query: {
        id: 'abc123'
      }
    });
  });
});
