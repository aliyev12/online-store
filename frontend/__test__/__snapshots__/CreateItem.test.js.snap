// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value="Item Title"
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={9998}
      />
    </label>
    <label
      htmlFor="title"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a Description"
        required={true}
        value="Item Description"
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
